controllers/istiorevisiontag/istiorevisiontag_controller.go

```diff
@@ -0,0 +1,565 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"istio.io/istio/pkg/ptr"
+)
+
+const (
+	revisionTagsChartName = "revisiontags"
+
+	sailOperatorReferencedRevisionLabel = "sailoperator.io/referenced-revision"
+)
+
+// Reconciler reconciles an IstioRevisionTag object
+type Reconciler struct {
+	client.Client
+	Scheme       *runtime.Scheme
+	Config       config.ReconcilerConfig
+	ChartManager *helm.ChartManager
+}
+
+func NewReconciler(reconcilerCfg config.ReconcilerConfig, client client.Client, scheme *runtime.Scheme, chartManager *helm.ChartManager) *Reconciler {
+	return &Reconciler{
+		Client:       client,
+		Scheme:       scheme,
+		Config:       reconcilerCfg,
+		ChartManager: chartManager,
+	}
+}
+
+// +kubebuilder:rbac:groups=sailoperator.io,resources=istiorevisiontags,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=sailoperator.io,resources=istiorevisiontags/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=sailoperator.io,resources=istiorevisiontags/finalizers,verbs=update
+// +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=mutatingwebhookconfigurations,verbs="*"
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *Reconciler) Reconcile(ctx context.Context, tag *v1alpha1.IstioRevisionTag) (ctrl.Result, error) {
+	log := logf.FromContext(ctx).WithValues("IstioRevisionTag", tag.Name)
+
+	rev, reconcileErr := r.doReconcile(ctx, tag)
+
+	log.Info("Reconciliation done. Updating labels and status.")
+	labelsErr := r.updateLabels(ctx, tag, rev)
+	statusErr := r.updateStatus(ctx, tag, rev, reconcileErr)
+
+	reconcileErr = errors.Unwrap(reconcileErr)
+
+	return ctrl.Result{}, errors.Join(reconcileErr, labelsErr, statusErr)
+}
+
+func (r *Reconciler) doReconcile(ctx context.Context, tag *v1alpha1.IstioRevisionTag) (*v1alpha1.IstioRevision, error) {
+	log := logf.FromContext(ctx).WithValues("IstioRevisionTag", tag.Name)
+	if err := r.validate(ctx, tag); err != nil {
+		return nil, err
+	}
+
+	log.Info("Retrieving referenced IstioRevision for IstioRevisionTag")
+	rev, err := r.getIstioRevision(ctx, tag.Spec.TargetRef)
+	if rev == nil || err != nil {
+		return nil, err
+	}
+
+	log.Info("Installing Helm chart")
+	return rev, r.installHelmCharts(ctx, tag, rev)
```

Comment:
That was indeed my first design, but that would leave us having to expose the `values.revisionTags` field in the IstioRevision spec, which I wanted to avoid because it'll be confusing and conflicting with the IstioRevisionTag CRD. Also, I figured that if we're going to have side-effects (ie 3rd party resources influencing the reconcile loop of a controller) it's probably better to keep them in a low-impact resource/chart like revision tags

Reply:
> Wouldn't that produce a simpler system that's closer to what upstream does?

I'm not sure - what upstream does is fetch the values of the existing control plane, re-render the entire istiod chart and hand-pick the MutatingWebhook YAML and apply it in isolation. So we're pretty close to what upstream does.

Reply:
> That was indeed my first design, but that would leave us having to expose the values.revisionTags field in the IstioRevision spec...

No, I didn't say "add `revisionTags` to `IstioRevision.spec.values`", but for the IstioRevision controller to set the `revisionTags` value only in the map that gets passed to the Helm render function.

> ...what upstream does is fetch the values of the existing control plane, re-render the entire istiod chart and hand-pick the MutatingWebhook YAML and apply it in isolation

OK, then this approach is indeed closer to upstream's, but I feel like it's a hack. The code that cherry-picks the MutatingWebhook YAML must know that the `revisionTags` value is only used in that YAML and in no other YAML produced by the istiod chart. So, it's somewhat fragile. 

I'm not against this approach, but I still feel that the alternative would be much simpler and more robust. It would also automatically cause both the old and the new revision to be reconciled whenever they become or stop being the active revision.